pykern/fconf.py

```diff
@@ -0,0 +1,569 @@
+        for f in files:
+            try:
+                self._add_file(f)
+            except Exception as e:
+                _exception_reason(e, f"fconf.Parser.file={f}")
+                raise
+        e = _Evaluator(parser=self)
+        for f in self.files.yml:
+            e.start(source=f)
+        self.result = e.global_fvars
+
+    def _add_file(self, path):
+        e = path.ext[1:].lower()
+        f = _File(
+            content=getattr(self, f"_ext_{e}")(path),
+            ext=e,
+            path=path,
+        )
+        if e in ("py", _BUILTINS_EXT):
+            self._add_macros(f)
+        elif e == "yml":
+            self._add_text_macros(f)
+        else:
+            raise ValueError(f"unhandled file ext={e}")
+        self.files.setdefault(e, []).append(f)
+
+    def _add_macro(self, macro):
+        n = macro.name
+        if n in self.macros:
+            raise ValueError(f"duplicate {macro}, other {self.macros[macro.name]}")
+        if n.startswith(_RESERVED_PREFIX) and macro.source.ext != _BUILTINS_EXT:
+            raise ValueError(f"macro={macro} may not begin with {_RESERVED_PREFIX}")
+        self.macros[macro.name] = macro
+
+    def _add_macros(self, source):
+        m = source.content
+        if not m:
+            return
+        for n, f in m.items():
+            self._add_macro(
+                _Macro(
+                    func=f,
+                    name=n,
+                    # TODO(robnagler) assert varnames don't begin with fconf_
+                    params=tuple(f.__code__.co_varnames),
+                    source=source,
+                ),
+            )
+
+    def _add_text_macros(self, source):
+        m = source.content.pkdel("fconf_macros")
+        if not m:
+            return
+        for n, c in m.items():
+            m = _TEXT_MACRO_DEF.search(n)
+            if not m:
+                raise ValueError(f"invalid macro name={n} {source}")
+            n = m.group(1)
+            # TODO: duplicate arg check
+            self._add_macro(
+                _YAMLMacro(
+                    content=c,
+                    name=n,
+                    # TODO(robnagler): check for dups
+                    # TODO(robnagler): assert varnames don't begin with fconf_
+                    params=tuple((x for x in _ARG_SEP.split(m.group(2)) if x)),
+                    source=source,
+                ),
+            )
+
+    def _ext_builtins(self, path):
+        return self._functions(_Builtins)
+
+    def _ext_py(self, path):
+        return self._functions(pykern.pkrunpy.run_path_as_module(path))
+
+    def _ext_yml(self, path):
+        return pykern.pkyaml.load_file(path)
+
+    def _functions(self, obj):
+        m = PKDict()
+        for n, o in inspect.getmembers(obj):
+            if inspect.isfunction(o):
+                m[n] = o
+        return m
+
+
+class _Builtins:
+    @staticmethod
+    def fconf_var(namespace, name):
+        return namespace._evaluator.fconf_var(name)
```

Comment:
`_evaluator` should be public if accessed like this

Reply:
It's not public, because I don't want clients of fconf to use it this way. This would be called "protected" in Java, but it's not really. All the classes except Parser are private. _Namespace is tricky because all macros are operating in this namespace so they can access its attributes. `__getattr__` is only called if an attribute is not found. I think this is good practice (_attrs can be accessed within a module).